gpus/example1/example1_using_cpp.cu

```diff
@@ -0,0 +1,298 @@
+#include <thrust/device_vector.h>
+#include <thrust/host_vector.h>
+#include <thrust/transform.h>
+
+template<typename T>
+using vector = thrust::host_vector<T>;
+
+// kernel function to perform vector c = alpha * a + b
+// size is the size of the vectors
+// __restrict__ tells the compiler a, b, and c will not
+// alias to each other (as in a != b, b != c, and a != c
+__global__ void saxpy(const float alpha, const float* __restrict__ a, const float* __restrict__ b, float* __restrict__ c, int size) {
+
+    // each block handles 1024 elements
+    // each block has 64 threads
+    // each thread in a block handles 4 of those elements
+    // we iterate 4 times to handle the 1024 elements
+    constexpr int elements_per_thread = 4;
+    
+    // threadIdx.x is the location within the grid
+    // blockIdx.x is the location of the block
+    // blockDim.x is the dimension of the block (threads per block)
+    // we know blockDim.x = 64 so we substitute 64 in
+    int tidx = threadIdx.x; 
+    
+    // we will block the vector addition with 4 elements per each thread
+   
+    int bidx = blockIdx.x;
+
+    const float* a_block = a + bidx * 1024;
+    const float* b_block = b + bidx * 1024;
+    float* c_block = c + bidx * 1024;
+
+    // if we know we are in bounds of the vector or if the vector size is evenly divisible by 1024
+    if(bidx < size / 1024 || (size + 1023) / 1024 == size / 1024) {
+
+        // we process 256 floating point calculations per loop
+        // float 512 floats loaded and 256 stored
+        #pragma unroll
+        for(int i = 0; i < 1024 / 64 / elements_per_thread; ++i) {
+            float4 reg_a = *(reinterpret_cast<const float4*>(a_block) + tidx + i * 64);
+            float4 reg_b = *(reinterpret_cast<const float4*>(b_block) + tidx + i * 64);
+            
+            reg_b.w += alpha * reg_a.w;
+            reg_b.x += alpha * reg_a.x;
+            reg_b.y += alpha * reg_a.y;
+            reg_b.z += alpha * reg_a.z;
+
+            *(reinterpret_cast<float4*>(c_block) + tidx + i * 64) = reg_b;
+        }
+    } else {
+        #pragma unroll
+        for(int i = 0; i < 1024 / 64 / elements_per_thread; ++i) {
+
+            // if the loads and stores are all in bounds proceed as normal
+            if(bidx + tidx + i * 256 + 3 < size) {
+                float4 reg_a = *(reinterpret_cast<const float4*>(a_block) + tidx + i * 64);
+                float4 reg_b = *(reinterpret_cast<const float4*>(b_block) + tidx + i * 64);
+                
+                reg_b.w += alpha * reg_a.w;
+                reg_b.x += alpha * reg_a.x;
+                reg_b.y += alpha * reg_a.y;
+                reg_b.z += alpha * reg_a.z;
+
+                *(reinterpret_cast<float4*>(c_block) + tidx + i * 64) = reg_b;
+            } else if(bidx + tidx + i * 256 < size) {
+
+                // perform each individually
+
+                float reg_a;
+                float reg_b;
+
+                for(int j = 0; j < 4; j++) {
```

Comment:
I read on Clang style guide once that pre-increment is more efficient. Don't really know how true it is, because I assume a good compiler should be able to optimise a post-increment in cases like this. Anyway, other loops in this file pre-increment their indices, so it's consistent to pre-increment here too.